hail/src/main/scala/is/hail/utils/prettyPrint/PrettyPrintWriter.scala

```diff
@@ -0,0 +1,187 @@
+          printNode(head, false)
+        }
+      }
+    }
+
+    // Process the top of kont until a non-empty ConcatK is found; move the first contained node
+    // to currentNode.
+    @tailrec def advance(): Unit = {
+      if (kont.isEmpty) {
+        currentNode = null
+      } else {
+        kont.peek() match {
+          case ConcatK(k) =>
+            if (k.isEmpty) {
+              kont.pop()
+              advance()
+            } else {
+              currentNode = k.next()
+            }
+          case PopGroupK =>
+            if (pendingOpens > 0) pendingOpens -= 1 else pendingCloses += 1
+            kont.pop()
+            advance()
+          case UnindentK(i) =>
+            indentation -= i
+            kont.pop()
+            advance()
+        }
+      }
+    }
+
+    def printNode(node: ScanedNode, flatten: Boolean): Unit = node match {
+      case TextN(t) =>
+        remainingInLine -= t.length
+        out.write(t)
+      case LineN(i, ifFlat: String) =>
+        if (flatten) {
+          remainingInLine -= ifFlat.length
+          out.write(ifFlat)
+        } else {
+          lines += 1
+          if (lines >= maxLines) throw new MaxLinesExceeded()
+          out.write('\n')
+          out.write(" " * i)
+          remainingInLine = math.min(width - i, ribbonWidth)
+        }
+      case GroupN(contents, start, stop) =>
+        val h = stop - start <= remainingInLine
+        var i = 0
+        while (i < contents.size) {
+          printNode(contents(i), h)
+          i += 1
+        }
+    }
+
+    def closeGroups(): Unit =
+      while (pendingCloses > 0) {
+        if (pendingGroups.isEmpty) {
+          pendingCloses = 0
+          return
+        }
+        val last = pendingGroups.removeLast()
+        last.end = globalPos
+        if (pendingGroups.isEmpty) {
+          printNode(last, true)
+        } else {
+          pendingGroups.getLast.contents += last
+        }
+        pendingCloses -= 1
+      }
+
+    def openGroups(): Unit = {
+      while (pendingOpens > 0) {
+        pendingGroups.addLast(GroupN(new ArrayBuilder[ScanedNode](), globalPos, -1))
+        pendingOpens -= 1
+      }
+    }
+
+    try {
+      while (currentNode != null) {
+        currentNode match {
+          case Text(t) =>
+            scan(TextN(t), t.length)
+            advance()
+          case Line(ifFlat) =>
+            closeGroups()
+            openGroups()
+            scan(LineN(indentation, ifFlat), ifFlat.length)
+            advance()
+          case Group(body) =>
+            kont.push(PopGroupK)
+            pendingOpens += 1
+            currentNode = body
+          case Indent(i, body) =>
+            indentation += i
+            kont.push(UnindentK(i))
+            currentNode = body
+          case Concat(bodyIt) =>
+            kont.push(ConcatK(bodyIt.iterator))
+            advance()
+        }
+      }
+      closeGroups()
+    } catch {
+      case _: MaxLinesExceeded =>
```

Comment:
why don't we want to throw here?

Reply:
I'm using the exception just as a control flow mechanism, it's not an error. This is to implement a `maxLines` option, which will keep formatting and printing lines until it hits `maxLines`, then stops without needing to traverse the rest of the input tree.

Reply:
So basically, it lets any of the helper functions break out of the outer while loop.

Reply:
`MaxLinesExceeded` should really be private to `prettyPrint`. Hopefully that makes the intent clearer. Adding a comment in the empty handler too.

Reply:
ah, got it!